wurst/event/DamageEvent.wurst

```diff
@@ -0,0 +1,238 @@
+        if firstListeners[prio] != null
+
+        if firstListeners[prio] == listener
+            firstListeners[prio] = listener.next
+            if listener.next == null and maxPriority == prio 
+                while firstListeners[prio] == null and prio > 0 
+                    prio--
+                maxPriority = prio 
+
+        else if listener.prev != null
+            listener.prev.next = listener.next
+
+        listener.next.prev = listener.prev
+
+
+    /* GETTERS */ 
+    static function getId() returns int  
+        return DamageInstance.current.id 
+
+    static function getSource() returns unit 
+        return DamageInstance.current.source
+
+    static function getTarget() returns unit 
+        return DamageInstance.current.target
+
+    static function getAmount() returns real 
+        return DamageInstance.current.amount
+
+    static function getOriginalAmount() returns real 
+        return DamageInstance.current.originalAmount
+
+    static function getType() returns DamageType 
+        return DamageInstance.current.damageType
+
+    static function getElement() returns DamageElement 
+        return DamageInstance.current.damageElement
+
+    static function isFiring() returns bool 
+        return DamageInstance.count > 0
+
+    /* SETTERS */
+    static function setAmount(real amount) 
+        DamageInstance.current.amount = amount
+
+    static function setNextDamageId(int id)
+        nextDamageId = id 
+
+    static function setNextDamageType(DamageType damageType)
+        nextDamageType = damageType 
+
+    static function setNextDamageElement(DamageElement damageElement)
+        nextDamageElement = damageElement
+
+    /* UTILS */
+    static function abortCurrent() 
+        abort = true 
+
+
+/* LISTENER */
+
+public abstract class DamageListener 
+    int priority = 0 
+    thistype prev = null 
+    thistype next = null
+
+    abstract function onEvent()
+
+    ondestroy
+        DamageEvent.removeListener(this)
+
+
+/* MISC */
+
+@configurable public function getDamageTypeName(DamageType t) returns string 
+    switch t 
+        case PHYSICAL
+            return "Physical"
+        case NATIVE_ABILITY
+            return "Ability"
+        case CODE
+            return "Ability"
+        default 
+            return "Unknown"
+
+@configurable public function getDamageDesignation(DamageType t, DamageElement e) returns string 
+    if t != DamageType.UNKNOWN and e != null
+        if t==DamageType.PHYSICAL and e==DAMAGE_ELEMENT_PHYSICAL
+            return "Pure Physical Damage"
+        return e.getName() + " " + getDamageTypeName(t) + " Damage"
+    else if t != DamageType.UNKNOWN
+        return getDamageTypeName(t) + " Damage"
+    else if e != null 
+        return e.getName() + " Damage"
+    return "Damage"
+
+
+/* INIT */
+
+public function initializeDamageEventSystem()
+    onUnitIndex( () -> EventListener.add(getIndexingUnit(), EVENT_UNIT_DAMAGED, () -> DamageEvent.onDamage()))
```

Comment:
Any particular reason for this?
Why not just use `EventListener.add(EVENT_UNIT_DAMAGED) ->`?